armory/scenarios/image_classification_load_adv_ds.py

```diff
@@ -0,0 +1,127 @@
+            logger.info(f"Applying internal {defense_type} defense to classifier")
+            classifier = load_defense_internal(config["defense"], classifier)
+
+        if model_config["fit"]:
+            classifier.set_learning_phase(True)
+            logger.info(
+                f"Fitting model {model_config['module']}.{model_config['name']}..."
+            )
+            fit_kwargs = model_config["fit_kwargs"]
+
+            logger.info(f"Loading train dataset {config['dataset']['name']}...")
+            train_data = load_dataset(
+                config["dataset"],
+                epochs=fit_kwargs["nb_epochs"],
+                split_type="train",
+                preprocessing_fn=preprocessing_fn,
+            )
+            if defense_type == "Trainer":
+                logger.info(f"Training with {defense_type} defense...")
+                defense = load_defense_wrapper(config["defense"], classifier)
+                defense.fit_generator(train_data, **fit_kwargs)
+            else:
+                classifier.fit_generator(train_data, **fit_kwargs)
+
+        if defense_type == "Transform":
+            # NOTE: Transform currently not supported
+            logger.info(f"Transforming classifier with {defense_type} defense...")
+            defense = load_defense_wrapper(config["defense"], classifier)
+            classifier = defense()
+
+        classifier.set_learning_phase(False)
+
+        # Evaluate the ART classifier on benign test examples
+        logger.info(f"Loading test dataset {config['dataset']['name']}...")
+        test_data_generator = load_dataset(
+            config["dataset"],
+            epochs=1,
+            split_type="test",
+            preprocessing_fn=preprocessing_fn,
+        )
+        logger.info("Running inference on benign examples...")
+        metrics_logger = metrics.MetricsLogger.from_config(config["metric"])
+
+        for x, y in tqdm(test_data_generator, desc="Benign"):
+            y_pred = classifier.predict(x)
+            metrics_logger.update_task(y, y_pred)
+        metrics_logger.log_task()
+
+        # Evaluate the ART classifier on adversarial test examples
+        logger.info("Generating / testing adversarial examples...")
+
+        ###################################################
+        # test_data_generator = load_dataset(
+        #     config["dataset"],
+        #     epochs=1,
+        #     split_type="test",
+        #     preprocessing_fn=preprocessing_fn,
+        # )
+
+        # for x, y in tqdm(test_data_generator, desc="Attack"):
+        #     x_adv = attack.generate(x=x)
+        #     y_pred_adv = classifier.predict(x_adv)
+        #     metrics_logger.update_task(y, y_pred_adv, adversarial=True)
+        #     metrics_logger.update_perturbation(x, x_adv)
+        ####################################################
+        attack = load_attack(config["attack"], classifier)
+
+        attack_type = config["attack"]["type"]
```

Comment:
My recommendation is to make a new attack class in `armory/art_experimental/attacks`, like `TransferAttack`, that takes dataset information as input. Then the dataset information here can be loaded as attack parameters into the attack config.

Reply:
I'm not sure creating a new type of attack is necessarily the simpler pathway, although I may not have thought about this from an implementation perspective as much as you have David.

Also, I think attack_type = "transfer" in the configuration file is not necessarily correct either.  The type of attack, given an adversarial dataset, depends on the model under attack, so it could be transfer or white-box.  Instead, the configuration should support a attack_type = "preloaded" (or something to that effect), to suggest that adversarial examples are not going to be generated on the fly in the scenario but rather from a file.

In the scenario, if attack_type = "preloaded", then an attack object will not be the instantiated; instead, the preloaded adversarial dataset will be loaded via a datagen and inference can be performed directly on that dataset.  On the other hand, if attack_type != "preloaded", then an attack object needs to be instantiated and used against a clean test dataset - of course, this will necessitate a change in the json format to allow two datasets to be declared under "dataset" field, one for the clean dataset and (optionally) one for the adversarial dataset.

Reply:
I like the idea of a "preloaded" attack_type, instead of a "transfer attack" (which might imply needing to load an additional model), with logic surrounding that attack_type. 

In terms of configuration, I think that it will be more understandable to have the dataset-specific information in the attack field instead of the dataset field - and this would only require a small (backwards-compatible) change to the configuration.

Another change to the logic in this case is that we'll need to provide paired clean and adversarial samples from the dataset in order to compute perturbation metrics for them.

Reply:
I agree on the need to pair.  I don't know if tfds has an API to pull data by the "key" argument.

Reply:
If your team can look into pairing / aligning datasets, that would be very helpful. This would solve an existing bug on the imagenet adversarial dataset (adversarial and clean splits don't follow the same ordering): https://github.com/twosixlabs/armory/issues/310